### ISSUE TITLE
derivable_impls false positive on nongeneric impl

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#7655]
```rust
pub struct Collection<T> {
    v: Vec<T>,
}

impl Default for Collection<String> {
    fn default() -> Self {
        Collection {
            v: Vec::new(),
        }
    }
}
```

```console
$ cargo clippy

warning: this `impl` can be derived
  --> src/main.rs:5:1
   |
5  | / impl Default for Collection<String> {
6  | |     fn default() -> Self {
7  | |         Collection {
8  | |             v: Vec::new(),
9  | |         }
10 | |     }
11 | | }
   | |_^
   |
   = note: `#[warn(clippy::derivable_impls)]` on by default
   = help: try annotating `Collection` with `#[derive(Default)]`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#derivable_impls
```

Clippy's suggestion is **not** equivalent to the handwritten impl. It is more generic which may not be desirable.

```rust
fn main() {
    let _ = Collection::default(); // works fine with handwritten Default
}
```

ambiguity with derived default but not handwritten default:

```console
error[E0282]: type annotations needed for `Collection<T>`
 --> src/main.rs:7:13
  |
7 |     let _ = Collection::default();
  |         -   ^^^^^^^^^^ cannot infer type for type parameter `T` declared on the struct `Collection`
  |         |
  |         consider giving this pattern the explicit type `Collection<T>`, where the type parameter `T` is specified
```

@HKalbasi @camsteffen

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.57.0-nightly (497ee321a 2021-09-09)
binary: rustc
commit-hash: 497ee321af3b8496eaccd7af7b437f18bab81abf
commit-date: 2021-09-09
host: x86_64-unknown-linux-gnu
release: 1.57.0-nightly
LLVM version: 13.0.0
```

**Comment by HKalbasi** [comment]
@camsteffen Shouldn't this line handle it?
```rust
if cx.tcx.type_of(item.def_id).definitely_has_param_types_or_consts(cx.tcx) {
```

**Comment by camsteffen** [comment]
I'm guessing that returns false because the type param is substituted with `String` and we need to go back to inspecting`type_of().kind()`.
